File: dotnet/src/Sodium/CSharp/Sodium.Frp.Tests/TimerTests.cs
using System;
using System.Collections.Generic;
using System.Threading;
using NUnit.Framework;
using Sodium.Frp.Time;
using Sodium.Functional;

namespace Sodium.Frp.Tests
{
    [TestFixture]
    public class TimerTests
    {
        [Test]
        public void SimultaneousTimerEvents()
        {
            TimerSystem<DateTime> ts = new SystemClockTimerSystem(e => { });
            Behavior<DateTime> time = ts.Time;
            List<DateTime> l = new List<DateTime>();
            Transaction.RunVoid(
                () =>
                {
                    DateTime now = time.Sample();
                    Stream<DateTime> a1 = ts.At(Cell.Constant(Maybe.Some(now.AddMilliseconds(99))));
                    Stream<DateTime> a2 = ts.At(Cell.Constant(Maybe.Some(now.AddMilliseconds(100))));
                    Stream<DateTime> a3 = ts.At(Cell.Constant(Maybe.Some(now.AddMilliseconds(100))));
                    Stream<DateTime> m = a1.OrElse(a2).OrElse(a3);
                    m.Listen(l.Add);
                });
            Thread.Sleep(200);
            Assert.That(l.Count, Is.EqualTo(2));
        }
    }
}

File: dotnet/src/Sodium/FSharp/Sodium.FSharp.Frp.Tests/TimerTests.fs
module Sodium.Frp.Tests.Timer

open System
open System.Collections.Generic
open NUnit.Framework
open Sodium.Frp
open System.Threading
open Sodium.Frp.Time

[<TestFixture>]
type ``Timer Tests``() =

    [<Test>]
    member __.``Simultaneous Timer Events``() =
        let ts = SystemClockTimerSystem(fun e -> ()) :> ITimerSystem<DateTime>
        let time = ts.Time
        let l = List<DateTime>()
        Transaction.run(fun () ->
            let now = time |> Behavior.sample
            let a1 = ts.At(Cell.constant(Some(now.AddMilliseconds(99.0))))
            let a2 = ts.At(Cell.constant(Some(now.AddMilliseconds(100.0))))
            let a3 = ts.At(Cell.constant(Some(now.AddMilliseconds(100.0))))
            let m = Stream.orElseAll [a1;a2;a3]
            m |> Stream.listen l.Add |> ignore)
        Thread.Sleep 200
        Assert.That(l.Count, Is.EqualTo(2))